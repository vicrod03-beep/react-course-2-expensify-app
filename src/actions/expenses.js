import { v4 as uuidv4 } from 'uuid'
import database from '../firebase/firebase'

// component calls actions generator
// action generator returns object
// component dispatches object
// redux store changes 


// component calls actions generator
// action generator returns function
// component dispatches function (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)


// add_expense
const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

//startAddExpense
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData
        const expense = { description, note, amount, createdAt }
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

// remove_expense
const removeExpense = ({
    id
} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

// edit_expense
const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export { addExpense, removeExpense, editExpense }