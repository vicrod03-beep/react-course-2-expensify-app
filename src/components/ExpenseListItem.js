import React from 'react'
import { NavLink } from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
    return (
        <div>
            <NavLink to={`/edit/${id}`} activeClassName="is-active">
                <h3>{description}</h3>
            </NavLink>
            <p>
                {numeral(amount / 100).format('$0,0.00')} 
                - 
                {moment(createdAt).format('MMMM Do, YYYY')}
            </p>
        </div>
    )
}

export default ExpenseListItem