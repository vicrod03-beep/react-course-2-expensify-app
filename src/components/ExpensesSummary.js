import React from 'react'
import { connect } from "react-redux"
import expensesTotal from '../selectors/expenses-total'
import getVisibleExpenses from '../selectors/expenses'
import numeral from 'numeral'

export const ExpensesSummary = (props) => (
    <div>
        {
            props.expensesCount === 0 ? '' :
            <h3> 
                Viewing {props.expensesCount} expense{props.expensesCount > 1 ? 's' : ''}, totalling {numeral(props.expensesSummary / 100).format('$0,0.00')}
            </h3>
        }
    </div>
)

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)
    return {
        expensesCount: visibleExpenses.length,
        expensesSummary: expensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)