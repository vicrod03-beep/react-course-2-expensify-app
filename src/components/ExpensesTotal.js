import React from 'react'

export default (props) => {
    return props.expenses.reduce((acc, { amount }, arr) => {
        return acc + amount
    }, 0)
}
 