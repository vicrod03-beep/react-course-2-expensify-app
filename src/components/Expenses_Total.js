import React from 'react'
import { connect } from "react-redux"
import getVisibleExpenses from "../selectors/expenses"

export const Expenses_Total = (props) => (
    <h3> Viewing {props.expenses.length} expenses, totalling {props.expenses.reduce((acc, { amount }, array) => {
                                                    return acc + amount
                                                    }, 0)}
    </h3>
)

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(Expenses_Total)