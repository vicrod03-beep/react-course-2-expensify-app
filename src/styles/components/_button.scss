.button {
    background: $dark-blue; 
    border: none;
    color: $light-white;
    display: inline-block;
    font-size: $font-size-large;
    font-weight: 300;
    line-height: 1;
    padding: $s-size;  
    text-decoration: none;
}

.button--link {
    background: none;
    border: none;
    color: $light-white;
    padding: 0
}

.button--secondary {
    background: #888888;
}