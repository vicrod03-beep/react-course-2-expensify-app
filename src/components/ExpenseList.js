import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'

export class ExpenseList extends React.Component {
    expenseLenght = () => this.props.expenses.length === 0 ? false : true
    render () {
        return (
            <div>
                {
                    this.expenseLenght() && 
                    this.props.expenses.map(expense => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                }
            </div>
        )
    }
}

/* export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? 
                <p>No expenses</p>
            : 
            props.expenses.map(expense => {
                return <ExpenseListItem key={expense.id} {...expense} />
            })
        }  
        <div>
            {props.expenses.length === 0 ? 
             '' :
            <ExpensesTotal expenses={props.expenses} />
            }
        </div>
    </div>
)
 */
const mapStateToProps = (state) => ({
        expenses: getVisibleExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)

