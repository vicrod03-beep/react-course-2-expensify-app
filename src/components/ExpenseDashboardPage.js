import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import Expenses_Total from './Expenses_Total'

const ExpenseDashboardPage = () => (
    <div>
        <Expenses_Total />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default ExpenseDashboardPage 