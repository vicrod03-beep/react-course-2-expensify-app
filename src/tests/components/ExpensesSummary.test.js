import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'
import expenses from '../fixtures/expenses'

let wrapper

beforeEach(() => {
    wrapper = shallow(<ExpensesSummary expensesCount={2} expensesSummary={94} />)
})

test('should view 2 expenses totalling 94', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should view 1 expense totalling 90', () => {
    wrapper.setProps({
        expensesCount: 1,
        expensesSummary: 90

    })
    expect(wrapper).toMatchSnapshot()
})