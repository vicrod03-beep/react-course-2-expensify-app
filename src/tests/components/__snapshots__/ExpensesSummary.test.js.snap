// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should view 1 expense totalling 90 1`] = `
<div>
  <h3>
    Viewing 
    1
     expense
    , totalling 
    $0.90
  </h3>
</div>
`;

exports[`should view 2 expenses totalling 94 1`] = `
<div>
  <h3>
    Viewing 
    2
     expense
    s
    , totalling 
    $0.94
  </h3>
</div>
`;
